docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+
+## Versioning of contracts
+Contracts are described an integer version number. A higher version number is not more recent, it just means different. In order to avoid conflicts, all contracts should be documented in the main branch of the dotnet repository with a version number which does not conflict with any other. It is expected that every version of every contract describes the same functionality/data layout/set of global values.
+
+## Contract data model
+Logically a contract may refer to another contract. If it does so, it will typically refer to other contracts by names which do not include the contract version. This is to allow for version flexibility. Logically once the contract data stream is fully processed, there is a single list of contracts that represents the set of contracts useable with whatever runtime instance is being processed.
+
+## Types of contracts
+
+There are 3 different types of contracts each representing a different phase of execution of the data contract system.
+
+### Composition contracts
+These contracts indicate the version numbers of other contracts. This is done to reduce the size of contract list needed in the runtime data stream. In general it is intended that as a runtime nears shipping, the product team can gather up all of the current versions of the contracts into a single magic value, which can be used to initialize most of the contract versions of the data contract system. A specific version number in the contract data stream for a given contract will override any composition contracts specified in the contract data stream. If there are multiple composition contracts in a stream which specify the same contract to have a different version, the first composition contract linearly in the runtime data stream wins. This is intended to allow for a composite contract for the architecture/os indepedent work, and a separate composite contract for the non independent work. If a contract is specified explicitly in the Data Contract Stream and a different version is specified via the composition contract mechanism, the explicitly specified contract takes precedence.
+
+### Fixed value contracts
+These contracts represent data which is entirely determined by the contract version + contract name. There are 2 subtypes of this form of contract.
+
+#### Global Value Contract
+A global value contract specifies numbers which can be referred to by other contracts. If a global value is specified directly in the Contract Data Stream, then the global value defintion in the Contract Data Stream takes precedence. The intention is that these global variable contracts represent magic numbers and values which are useful for the operation of algorithmic contracts. For instance, we will likely have a `TargetPointerSize` global value represented via a contract, and things like `FEATURE_SUPPORTS_COM` can also be a global value contract, with a value of 1.
```

Comment:
nit: I'm not sure `TargetPointerSize` is a great example.  Or at least it's a special example because I think that is a contract whose information is probably derived from the descriptor header or even /a priori/ known to the data contract reader when it connects to a target - it's hard to do any truly portable parsing without knowing the pointer size.

Reply:
Yeah, the 2 cases I found for this are basically TargetPointerSize and feature compilation flags. Not sure if that's worth a whole mechanism, or if we could encode stuff in a bitfield somewhere.

Reply:
I think it may be a little arbitrary what gets defined as globals vs what are defined as parameterized constants attached to contracts. Ignoring how we categorize them though I'd also expect we'll may want to encode values specified in C++ constants and C++ enums. For example there has been some shuffling of MethodTable flag values in recent versions so we might want a way to encode those values in case there are any further changes in the meaning we assign to different bit positions.

Reply:
> we'll may want to encode values specified in C++ constants and C++ enums. 

Agreed.  One thing that Mono does in a bunch of places is to conditionally define enums based on compile-time runtime feature flags.  For example the mini IR has platform specific opcodes that are only defined on that platform.  (we will not need to look at IR opcodes, but there are other enums with conditional definitions).

I _think_ that sort of situation can be captured by global value contracts.